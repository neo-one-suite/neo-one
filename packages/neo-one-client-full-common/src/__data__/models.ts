import { common, ContractParameterTypeModel, ECPoint, SignatureString, UInt160 } from '@neo-one/client-common';
import { constants } from '@neo-one/utils';
import {
  ContractABIModel,
  ContractEventDescriptorModel,
  ContractFeaturesModel,
  ContractGroupModel,
  ContractManifestModel,
  ContractMethodDescriptorModel,
  ContractParameterDefinitionModel,
  ContractPermissionDescriptorModel,
  ContractPermissionModel,
  ContractStateModel,
} from '../models';

export const contractParamDefinitionModel = {
  any: new ContractParameterDefinitionModel({
    type: ContractParameterTypeModel.Any,
    name: 'param',
  }),
  boolean: new ContractParameterDefinitionModel({
    type: ContractParameterTypeModel.Boolean,
    name: 'param',
  }),
  byteArray: new ContractParameterDefinitionModel({
    type: ContractParameterTypeModel.ByteArray,
    name: 'param',
  }),
  hash160: new ContractParameterDefinitionModel({
    type: ContractParameterTypeModel.Hash160,
    name: 'param',
  }),
  hash256: new ContractParameterDefinitionModel({
    type: ContractParameterTypeModel.Hash256,
    name: 'param',
  }),
  array: new ContractParameterDefinitionModel({
    type: ContractParameterTypeModel.Array,
    name: 'param',
  }),
  integer: new ContractParameterDefinitionModel({
    type: ContractParameterTypeModel.Integer,
    name: 'param',
  }),
  interopInterface: new ContractParameterDefinitionModel({
    type: ContractParameterTypeModel.InteropInterface,
    name: 'param',
  }),
  map: new ContractParameterDefinitionModel({
    type: ContractParameterTypeModel.Map,
    name: 'param',
  }),
  publicKey: new ContractParameterDefinitionModel({
    type: ContractParameterTypeModel.PublicKey,
    name: 'param',
  }),
  signature: new ContractParameterDefinitionModel({
    type: ContractParameterTypeModel.Signature,
    name: 'param',
  }),
  string: new ContractParameterDefinitionModel({
    type: ContractParameterTypeModel.String,
    name: 'param',
  }),
  void: new ContractParameterDefinitionModel({
    type: ContractParameterTypeModel.Void,
    name: 'param',
  }),
};

export const contractMethodDescriptorModel = (
  parameters: readonly ContractParameterDefinitionModel[] = [contractParamDefinitionModel.boolean],
  returnType: ContractParameterTypeModel = ContractParameterTypeModel.Void,
  name = 'function',
  offset = 0,
) => new ContractMethodDescriptorModel({ name, parameters, returnType, offset });

export const contractEventDescriptorModel = (
  parameters: readonly ContractParameterDefinitionModel[] = [contractParamDefinitionModel.boolean],
  name = 'event',
) => new ContractEventDescriptorModel({ name, parameters });

export const contractAbiModel = (
  methods: readonly ContractMethodDescriptorModel[] = [contractMethodDescriptorModel()],
  events: readonly ContractEventDescriptorModel[] = [contractEventDescriptorModel()],
  hash = common.bufferToUInt160(Buffer.alloc(20, 1)),
) => new ContractABIModel({ hash, methods, events });

export const contractGroupModel = (
  publicKey: ECPoint = common.stringToECPoint(constants.PRIVATE_NET_PUBLIC_KEY),
  signature: Buffer = Buffer.from(
    'ccaab040cc25021c91567b75db4778853441869157b8f6aad960cdcf1069812480027a528ca9b98e2205027de20696f848cf81824eeb7af1d5110870870ceb67',
    'hex',
  ),
) => new ContractGroupModel({ publicKey, signature });

export const contractPermissionDescriptorModel = (hashOrGroupType: 'uint160' | 'ecpoint' | undefined) => {
  let hashOrGroup: UInt160 | ECPoint | undefined;
  if (hashOrGroupType === 'uint160') {
    hashOrGroup = common.bufferToUInt160(Buffer.alloc(20, 1));
  } else if (hashOrGroupType === 'ecpoint') {
    hashOrGroup = common.stringToECPoint(constants.PRIVATE_NET_PUBLIC_KEY);
  }

  return new ContractPermissionDescriptorModel({ hashOrGroup });
};

export const contractPermissionModel = (
  hashOrGroupType: 'uint160' | 'ecpoint' | undefined,
  methods: readonly string[] = [],
) => new ContractPermissionModel({ contract: contractPermissionDescriptorModel(hashOrGroupType), methods });

export const contractManifestModel = (
  groups: readonly ContractGroupModel[] = [contractGroupModel()],
  features: ContractFeaturesModel = ContractFeaturesModel.HasStoragePayable,
  abi: ContractABIModel = contractAbiModel(),
  permissions: readonly ContractPermissionModel[] = [contractPermissionModel('uint160', ['method1'])],
  trusts: readonly UInt160[] = [common.bufferToUInt160(Buffer.alloc(20, 1))],
  safeMethods: readonly string[] = ['method1', 'method2'],
  supportedStandards: readonly string[] = [],
) => new ContractManifestModel({ groups, features, supportedStandards, abi, permissions, trusts, safeMethods });

export const contractModel = (
  id = 1,
  script: Buffer = Buffer.alloc(25),
  manifest: ContractManifestModel = contractManifestModel(),
) => new ContractStateModel({ id, script, manifest });
