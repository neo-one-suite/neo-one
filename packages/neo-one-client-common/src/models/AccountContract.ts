import _ from 'lodash';
import { common, ECPoint, UInt160 } from '../common';
import { crypto } from '../crypto';
import { scriptHashToAddress } from '../helpers';
import { JSONHelper } from '../JSONHelper';
import { utils } from '../utils';
import {
  ContractParameterTypeModel,
  toContractParameterType,
  toJSONContractParameterType,
} from './ContractParameterTypeModel';
import { AccountContractJSON } from './types';

export interface AccountContractAdd {
  readonly script: Buffer;
  readonly parameterList: readonly ContractParameterTypeModel[];
  readonly scriptHash?: UInt160;
  readonly address?: string;
}
export class AccountContract {
  public static createMultiSigContract(m: number, publicKeys: readonly ECPoint[]): AccountContract {
    return new AccountContract({
      script: crypto.createMultiSignatureRedeemScript(m, publicKeys),
      parameterList: _.range(m).map(() => ContractParameterTypeModel.Signature),
    });
  }

  public static createSignatureContract(publicKey: ECPoint): AccountContract {
    return new AccountContract({
      script: crypto.createSignatureRedeemScript(publicKey),
      parameterList: [ContractParameterTypeModel.Signature],
    });
  }

  public static fromJSON(json: AccountContractJSON): AccountContract {
    return new AccountContract({
      script: JSONHelper.readBase64Buffer(json.script),
      parameterList: json.parameterlist.map(toContractParameterType),
      scriptHash: JSONHelper.readUInt160(json.scripthash),
      address: json.address,
    });
  }

  public readonly script: Buffer;
  public readonly parameterList: readonly ContractParameterTypeModel[];
  public readonly scriptHashInternal: () => UInt160;
  public readonly addressInternal: () => string;

  public constructor({ script, parameterList, scriptHash: scriptHashIn, address: addressIn }: AccountContractAdd) {
    this.script = script;
    this.parameterList = parameterList;
    this.scriptHashInternal = scriptHashIn ? () => scriptHashIn : utils.lazy(() => crypto.hash160(this.script));
    this.addressInternal = addressIn
      ? () => addressIn
      : utils.lazy(() => scriptHashToAddress(common.uInt160ToString(this.scriptHashInternal())));
  }

  public get scriptHash() {
    return this.scriptHashInternal();
  }

  public get address() {
    return this.addressInternal();
  }

  public serializeJSON(): AccountContractJSON {
    return {
      script: JSONHelper.writeBase64Buffer(this.script),
      parameterlist: this.parameterList.map(toJSONContractParameterType),
      scripthash: JSONHelper.writeUInt160(this.scriptHash),
      address: this.address,
    };
  }
}
