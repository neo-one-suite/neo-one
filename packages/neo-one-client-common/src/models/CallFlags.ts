export enum CallFlags {
  None = 0,
  ReadStates = 0b00000001,
  WriteStates = 0b00000010,
  AllowCall = 0b00000100,
  AllowNotify = 0b00001000,
  // tslint:disable-next-line: no-bitwise
  States = ReadStates | WriteStates,
  // tslint:disable-next-line: no-bitwise
  ReadOnly = ReadStates | AllowCall,
  // tslint:disable-next-line: no-bitwise
  All = States | AllowCall | AllowNotify,
}
