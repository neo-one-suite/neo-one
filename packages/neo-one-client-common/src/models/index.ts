export * from './AccountContract';
export * from './ContractParameterTypeModel';
export * from './SignerModel';
export * from './VerifyResultModel';
export * from './Serializable';
export * from './WitnessModel';
export * from './WitnessScopeModel';
export * from './transaction';
export * from './nep17';
export * from './types';
export * from './vm';
export * from './trigger';
export * from './NefFileModel';
export * from './CallFlags';
export * from './MethodTokenModel';
