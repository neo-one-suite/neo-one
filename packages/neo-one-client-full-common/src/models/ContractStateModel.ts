import { NefFileModel, UInt160 } from '@neo-one/client-common';
import { ContractManifestModel } from './manifest';

export interface ContractStateModelAdd<
  TContractManifest extends ContractManifestModel = ContractManifestModel,
  TNefFile extends NefFileModel = NefFileModel
> {
  readonly id: number;
  readonly updateCounter: number;
  readonly hash: UInt160;
  readonly nef: TNefFile;
  readonly manifest: TContractManifest;
}

export class ContractStateModel<
  TContractManifest extends ContractManifestModel = ContractManifestModel,
  TNefFile extends NefFileModel = NefFileModel
> {
  public readonly id: number;
  public readonly updateCounter: number;
  public readonly hash: UInt160;
  public readonly nef: TNefFile;
  public readonly manifest: TContractManifest;

  public constructor({ nef, hash, manifest, id, updateCounter }: ContractStateModelAdd<TContractManifest, TNefFile>) {
    this.id = id;
    this.updateCounter = updateCounter;
    this.hash = hash;
    this.nef = nef;
    this.manifest = manifest;
  }
}
