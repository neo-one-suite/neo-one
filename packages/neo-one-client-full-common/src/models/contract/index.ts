export * from './ContractFeaturesModel';
export * from './ContractManifestModel';
export * from './ContractModel';
export * from './abi';
