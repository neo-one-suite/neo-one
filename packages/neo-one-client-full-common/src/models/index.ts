export * from './BaseState';
export * from './contract';
export * from './permissions';
