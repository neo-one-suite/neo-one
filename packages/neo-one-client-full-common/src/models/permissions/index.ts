export * from './ContractGroupModel';
export * from './ContractPermissionDescriptorModel';
export * from './ContractPermissionModel';
