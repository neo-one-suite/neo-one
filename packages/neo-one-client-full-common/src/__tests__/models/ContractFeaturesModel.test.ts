import { assertContractFeature, getContractProperties, HasPayable, HasStorage } from '../../models';

describe('ContractFeaturesModel', () => {
  test('isContractFeatures', () => {
    expect(assertContractFeature(0)).toEqual(0);
    expect(assertContractFeature(1)).toEqual(1);
    expect(assertContractFeature(4)).toEqual(4);
    expect(assertContractFeature(5)).toEqual(5);

    const assertThrowNeg1 = () => assertContractFeature(-1);
    expect(assertThrowNeg1).toThrowErrorMatchingSnapshot();
    const assertThrow2 = () => assertContractFeature(2);
    expect(assertThrow2).toThrowErrorMatchingSnapshot();
    const assertThrow3 = () => assertContractFeature(3);
    expect(assertThrow3).toThrowErrorMatchingSnapshot();
    const assertThrow6 = () => assertContractFeature(6);
    expect(assertThrow6).toThrowErrorMatchingSnapshot();
  });

  test('getContractProperties', () => {
    expect(getContractProperties({ hasStorage: false, payable: false })).toEqual(0);
    expect(getContractProperties({ hasStorage: false, payable: true })).toEqual(4);
    expect(getContractProperties({ hasStorage: true, payable: false })).toEqual(1);
    expect(getContractProperties({ hasStorage: true, payable: true })).toEqual(5);
  });

  test('HasStorage', () => {
    expect(HasStorage.has(0)).toEqual(false);
    expect(HasStorage.has(1)).toEqual(true);
    expect(HasStorage.has(2)).toEqual(false);
    expect(HasStorage.has(3)).toEqual(false);
    expect(HasStorage.has(4)).toEqual(false);
    expect(HasStorage.has(5)).toEqual(true);
    expect(HasStorage.has(6)).toEqual(false);
  });

  test('HasPayable', () => {
    expect(HasPayable.has(0)).toEqual(false);
    expect(HasPayable.has(1)).toEqual(false);
    expect(HasPayable.has(2)).toEqual(false);
    expect(HasPayable.has(3)).toEqual(false);
    expect(HasPayable.has(4)).toEqual(true);
    expect(HasPayable.has(5)).toEqual(true);
    expect(HasPayable.has(6)).toEqual(false);
  });
});
