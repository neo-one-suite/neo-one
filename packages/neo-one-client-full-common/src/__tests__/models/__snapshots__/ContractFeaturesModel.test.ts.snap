// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractFeaturesModel isContractFeatures 1`] = `"Expected contract feature, found: -1"`;

exports[`ContractFeaturesModel isContractFeatures 2`] = `"Expected contract feature, found: 2"`;

exports[`ContractFeaturesModel isContractFeatures 3`] = `"Expected contract feature, found: 3"`;

exports[`ContractFeaturesModel isContractFeatures 4`] = `"Expected contract feature, found: 6"`;
